### ISSUE TITLE
Incorrect output dimensions with dilated convolution and valid padding

### ISSUE THREAD

**Issue by yamiyukiharu** [onnx/tensorflow-onnx#945]
**Describe the bug**
Upon converting my Resnet50 based model from tensorflow frozen graph to onnx, then executing it with trtexec, I get the following error:

> [05/30/2020-23:52:33] [E] [TRT] resnet_v1_50/block4/unit_1/bottleneck_v1/add: elementwise inputs must have same dimensions or follow broadcast rules (input dimensions were [1,2048,45,80] and [1,2048,46,81]).
> [05/30/2020-23:52:33] [E] [TRT] resnet_v1_50/block4/unit_1/bottleneck_v1/add: elementwise inputs must have same dimensions or follow broadcast rules (input dimensions were [1,2048,45,80] and [1,2048,46,81]).
> [05/30/2020-23:52:33] [E] [TRT] resnet_v1_50/block4/unit_1/bottleneck_v1/add: elementwise inputs must have same dimensions or follow broadcast rules (input dimensions were [1,2048,45,80] and [1,2048,46,81]).
> [05/30/2020-23:52:33] [E] [TRT] resnet_v1_50/block4/unit_1/bottleneck_v1/add: elementwise inputs must have same dimensions or follow broadcast rules (input dimensions were [1,2048,45,80] and [1,2048,46,81]).
> [05/30/2020-23:52:33] [E] [TRT] resnet_v1_50/block4/unit_1/bottleneck_v1/add: elementwise inputs must have same dimensions or follow broadcast rules (input dimensions were [1,2048,45,80] and [1,2048,46,81]).
> While parsing node number 96 [Conv]:
> ERROR: builtin_op_importers.cpp:461 In function importConv:
> [8] Assertion failed: nbSpatialDims == kernel_weights.shape.nbDims - 2
> [05/30/2020-23:52:33] [E] Failed to parse onnx file
> [05/30/2020-23:52:33] [E] Parsing model failed
> [05/30/2020-23:52:33] [E] Engine creation failed
> [05/30/2020-23:52:33] [E] Engine set up failed
> ****

This only happens when I specify the input dimensions, it runs perfectly when I make it an input with dynamic shape. Hence I think it is a problem with the tf2onnx conversion. 
Below is a screenshot of the problematic layer in the original frozen graph:
![Screenshot from 2020-05-31 00-07-43](https://user-images.githubusercontent.com/12526810/83333383-c8910300-a2d2-11ea-8cf3-cfb1dc0c5ab3.png)

Below is a screenshot of the problematic layer in the converted onnx file:

![Screenshot from 2020-05-31 00-05-06](https://user-images.githubusercontent.com/12526810/83333306-75b74b80-a2d2-11ea-8827-25b0ab720582.png)

The original graph from Tensorflow has VALID padding, so the shapes should always match right?

**Urgency**
Important for project

**System information**
- OS Platform and Distribution: Ubuntu 18.04
- Tensorflow Version: 1.14
- Python version: 3.6

**Comment by jignparm** [comment]
Is it easy to share a link to the TF model?  The shapes for the convolution operators look compatible above (between the TF and Onnx diagrams). The ordering of the dimensions will not be the same, which is expected, since TF is NHWC and Onnx is NCHW.

A quick validation of the Onnx model would be to load it using OnnxRuntime - if it loads/runs using OnnxRuntime, it's most likely a TRT error.

**Comment by jignparm** [comment]
Any link to the model yet? Would be helpful to debug.

**Comment by phager90** [comment]
@jignparm:
I noticed something before in the padding computation that could be related to this issue:

https://github.com/onnx/tensorflow-onnx/blob/master/tf2onnx/onnx_opset/nn.py L240
```
# Calculate paddings.
        for i in range(spatial):
            pad = (
                (output_shape[i + 2] - 1) * strides[i]
                + dilations[i] * kernel_shape[i]
                - input_shape[i + 2]
            )
```

If you compare it with the pytorch input/output relation formula:
https://pytorch.org/docs/master/generated/torch.nn.Conv2d.html

Could it be that the dilation term should be instead:
`                + dilations[i] * (kernel_shape[i] - 1) + 1`

If this is the case, this would mean that the padding is incorrectly computed in the layer with dilation > 1 AND stride >1.

**Comment by jignparm** [comment]
1. Does making the change above cause your model to pass and run in TRT (you can edit the file manually to test)? If yes, then it may be a real bug, and require a fix in tf2onnx.

1. Can you try running your model using [OnnxRuntime](https://github.com/microsoft/onnxruntime/tree/master/samples#Python) as well? If the model(s) run under OnnxRuntime, it  means tf2onnx is fine, and TRT may require a patch.

These steps should help isolate the problem down to TRT or else Tf2onnx. This code section above is working for several convolution models, so it seems much quicker to run the 2 steps above than to look at the padding formula. Also, if there's a link to the model, it'll help in debugging the issue (if it turns out to be tf2onnx issue).

**Comment by TomWildenhain-Microsoft** [comment]
I can confirm that there is a bug with padding and am working on a fix.

**Comment by TomWildenhain-Microsoft** [comment]
@yamiyukiharu I made a change but I don't know if it fixed your bug.  Let me know if this is still an issue.

**Comment by guschmue** [comment]
we assume this fixed the issue.
